python/lsst/obs/base/read_defects.py

```diff
@@ -0,0 +1,40 @@
+from lsst.meas.algorithms import Defects
+import os
+import glob
+from dateutil import parser
+
+
+def read_defects_one_chip(root, chip_name, camera):
+    files = glob.glob(os.path.join(root, chip_name.lower(), '*.ecsv'))
+    parts = os.path.split(root)
+    instrument = os.path.split(parts[0])[1]  # convention is that these reside at <instrument>/defects
+    defect_dict = {}
+    for f in files:
+        date_str = os.path.splitext(os.path.basename(f))[0]
+        valid_start = parser.parse(date_str)
+        defect_dict[valid_start] = Defects.readText(f)
+        check_metadata(defect_dict[valid_start], valid_start, instrument, camera[chip_name].getId())
+    return defect_dict
+
+
+def check_metadata(defects, valid_start, instrument, chip_id):
+    md = defects.getMetadata()
+    finst = md.get('INSTRUME')
+    fchip_id = md.get('DETECTOR')
+    fcalib_date = md.get('CALIBDATE')
+    if not (finst, int(fchip_id), fcalib_date) == (instrument, chip_id, valid_start.isoformat()):
+        raise ValueError("Path and file metadata do not agree:\n" +
+                         "Path metadata: %s, %s, %s\n"%(instrument, chip_id, valid_start.isoformat()) +
+                         "File metadata: %s, %s, %s\n"%(finst, fchip_id, fcalib_date))
+
+
+def read_all_defects(root, camera):
+    dirs = os.listdir(root)  # assumes all directories contain defects
+    dirs = [d for d in dirs if os.path.isdir(os.path.join(root, d))]
+    defects_by_chip = {}
+    name_map = {det.getName().lower(): det.getName() for
+                det in camera}  # we assume the directories have been lowered
```

Comment:
I find the comment cryptic. I assume you are saying you assume the directory names are lowercase, but it's not immediately clear how that relates to the code in question.

I suggest that instead of assuming directory names are lowercase, you do one of two things:
* Check that dir names are lowercase and raise if not. One logical place to do this is right after `dirs = os.listdir(root)`.
*  Replace `chip_name = name_map[os.path.basename(d)]` with `chip_name = name_map[os.path.basename(d).lower()]`. This is simple but will not catch the case of two directories whose names only differs by case.

Reply:
The problem is that the names are not always lower case in the `Camera` so we have to map between on disk names and what the camera thinks.  We wanted to assume the filesystem was lower to avoid issues with non-case sensitive file systems.

I've moved all the camera translation in to `read_all_defects`.  Maybe that is clearer?

Reply:
I was suggesting that when you compare names you lowercase them (and I suggested two different ways to do that). I _think_ that avoids the need for this dict and simplifies the code a bit.